write-fonts/src/tables/meta.rs

```diff
@@ -0,0 +1,137 @@
+//! The [meta (Metadata)](https://docs.microsoft.com/en-us/typography/opentype/spec/meta) table
+
+use std::fmt::Display;
+
+include!("../../generated/generated_meta.rs");
+
+pub const DLNG: Tag = Tag::new(b"dlng");
+pub const SLNG: Tag = Tag::new(b"slng");
+
+/// Metadata in the `meta` table, associated with some tag.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Metadata {
+    /// For the 'dlng' and 'slng' tags
+    ScriptLangTags(Vec<ScriptLangTag>),
+    /// For other tags
+    Other(Vec<u8>),
+}
+
+/// A ['ScriptLangTag'] value.
+///
+/// This is currently just a string and we do not perform any validation,
+/// but we should do that (TK open issue)
+///
+/// [`ScriptLangTag`]: https://learn.microsoft.com/en-us/typography/opentype/spec/meta#scriptlangtag-values
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ScriptLangTag(String);
+
+/// An error for if a [`ScriptLangTag`] does not conform to the specification.
+#[derive(Clone, Debug)]
+#[non_exhaustive] // so we can flesh this out later without breaking anything
+pub struct InvalidScriptLangTag;
+
+impl ScriptLangTag {
+    pub fn new(raw: String) -> Result<Self, InvalidScriptLangTag> {
+        Ok(Self(raw))
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl Display for InvalidScriptLangTag {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("ScriptLangTag was malformed")
+    }
+}
+
+impl std::error::Error for InvalidScriptLangTag {}
+
+impl DataMapRecord {
+    fn validate_data_type(&self, ctx: &mut ValidationCtx) {
+        if matches!(
+            (self.tag, self.data.as_ref()),
+            (SLNG | DLNG, Metadata::Other(_))
+        ) {
+            ctx.report("'slng' or 'dlng' tags use ScriptLangTag data");
+        }
+    }
+}
+
+impl FontWrite for Metadata {
+    fn write_into(&self, writer: &mut TableWriter) {
+        let len = match self {
+            Metadata::ScriptLangTags(langs) => {
+                let mut len = 0;
+                for lang in langs {
+                    if len > 0 {
+                        b','.write_into(writer);
+                        len += 1;
+                    }
+                    lang.0.as_bytes().write_into(writer);
+                    len += lang.0.as_bytes().len();
+                }
+                len
+            }
+            Metadata::Other(vec) => {
+                vec.write_into(writer);
+                vec.len()
+            }
+        };
+
+        let len: u32 = len.try_into().unwrap();
```

Comment:
I don't love the unwrap, is this common in write-fonts? Should file an issue to make write_into fallible and remove that?

Reply:
By design `FontWrite` is infallible; we do all the error checking in an earlier validation pass.

In generated code, we do add a check that an array's length can be represented by the type that stores the length, but I didn't bother hand-writing that here, given that any single font table's length must be representable in 32 bits or else its offset cannot be stored in the root table directory.

Reply:
Maybe we should comment as much and do an infallible production of len?

Reply:
So you've brought up something that is ultimately a larger philosophical question that we should discuss, which is what we want to do when the user really wants to write nonsensical table data.

Some possible options:

- let them
- let them, but with hand guards: for instance always truncating collections to their maximum allowed length
- crash

bear in mind that we do already return an error if they fail validation, so this decision only really concerns the case where,

- someone is manually skipping validation (which would require very intentional work, validation happens automatically if you use `dump_table` so you'd need to basically reimplement that function yourself)
- we have failed to add some validation check (this is definitely going to be the case _somewhere_)

so my personal preference right now would be to crash, because it tells us in this latter case very clearly that we need to add a validation check for something, but maybe this is not appropriate for all cases?

In this specific case, I'll add the validation check manually and leave the unwrap, since that is consistent with what we do elsewhere. The only way you could hit the unwrap is if you manually write your own serialization code that skips validation, and then construct a completely unreasonable `meta` table that you pass to your custom serialization code, so it's not going to be hit by accident...